### PR TITLE
mark fork as unsafe

### PR DESCRIPTION
**Pull request by snsmac**
Reasoning: After a fork() in a multithreaded program, only async-signal-safe functions may be called by the child.
Calling everything else (including malloc(), mutexes) is undefined behavior.

This is obviously a breaking change 

References:
[std discussion](https://github.com/rust-lang/rust/issues/39575)
[std merge request](https://github.com/rust-lang/rust/pull/58059)
[signal-safety(7)](https://manpages.debian.org/testing/manpages/signal-safety.7.en.html)
[fork(2)](http://pubs.opengroup.org/onlinepubs/9699919799/functions/fork.html)

Closes: https://github.com/nix-rust/nix/issues/1030

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -41,7 +41,8 @@ This project adheres to [Semantic Versioning](http://semver.org/).
   ([#1020](https://github.com/nix-rust/nix/pull/1020))
 - Replaced `CmsgSpace` with the `cmsg_space` macro.
   ([#1020](https://github.com/nix-rust/nix/pull/1020))
-
+- Marked `fork` as unsafe function
+  ([#1020](https://github.com/nix-rust/nix/pull/1020))
```
Comment (asomers) [inline CHANGELOG.md]: The PR number is wrong.

File: src/unistd.rs [hunk 2]
```diff
@@ -192,7 +192,7 @@ impl ForkResult {
 /// ```no_run
 /// use nix::unistd::{fork, ForkResult};
 ///
-/// match fork() {
+/// match unsafe {fork()} {
```
Comment (asomers) [inline src/unistd.rs]: You should add a section to the docs describing how to use `fork` safely, like this:

```
/// # Safety
/// 
/// To use this function safely, you must do A, B, and C.
```

File: src/unistd.rs [hunk 3]
```diff
@@ -210,17 +215,22 @@ impl ForkResult {
 /// I'm a new child process
 /// ```
 ///
+///
 /// # Safety
+/// Make sure to read the [man page].
+/// Fork has some properties that can lead to unintended behavior.
+/// Especially the handling of mutexes and open files can be tricky.
```
Comment (asomers) [inline src/unistd.rs]: sentence fragment.

File: src/unistd.rs [hunk 4]
```diff
@@ -210,17 +215,22 @@ impl ForkResult {
...
 ///
 /// In a multithreaded program, only [async-signal-safe] functions like `pause`
-/// and `_exit` may be called by the child (the parent isn't restricted). Note
-/// that memory allocation may **not** be async-signal-safe and thus must be
+/// and `_exit` are safe to call by the child until it calls `execve`(the parent isn't restricted).
```
Comment (asomers) [inline src/unistd.rs]: Technically, there are other functions that can take the place of `execve`, like `fexecve`.  Better just to say "until it execs another program" or something like that.

File: src/unistd.rs [hunk 5]
```diff
@@ -210,17 +215,22 @@ impl ForkResult {
...
+/// Note that memory allocation may **not** be async-signal-safe and thus must be
 /// prevented.
 ///
 /// Those functions are only a small subset of your operating system's API, so
 /// special care must be taken to only invoke code you can control and audit.
 ///
 /// [async-signal-safe]: http://man7.org/linux/man-pages/man7/signal-safety.7.html
+/// [man page]: http://man7.org/linux/man-pages/man2/fork.2.html
```
Comment (asomers) [inline src/unistd.rs]: For standardized functions, it's better to use the open group's man page: http://pubs.opengroup.org/onlinepubs/9699919799/functions/fork.html

File: src/unistd.rs [hunk 6]
```diff
@@ -210,21 +215,26 @@ impl ForkResult {
 /// I'm a new child process
 /// ```
 ///
+///
 /// # Safety
+/// Make sure to read the [man page].
```
Comment (asomers) [inline src/unistd.rs]: I consider that RTFM goes without saying.  You can remove this sentence.

**Comment by dequis** [comment]
Why did this get closed?

**Comment by vi** [comment]
Shall this pull request be resubmitted with the documentation changes proposed above addressed?

**Comment by asomers** [comment]
Yeah, it should.

**Pull request closed without merging**